File: engineering/engineering/report/hourly_downtime_report/hourly_downtime_report.js
// Copyright (c) 2026, Isambane Mining (Pty) Ltd
// For license information, please see license.txt

frappe.query_reports["Hourly Downtime Report"] = {
    filters: [
        {
            fieldname: "report_date",
            label: __("Date"),
            fieldtype: "Date",
            reqd: 1,
            default: frappe.datetime.get_today(),
            onchange: function () {
                frappe.query_report.refresh();
            }
        },
        {
            fieldname: "hour_slot",
            label: __("Hour"),
            fieldtype: "Select",
            reqd: 1,
            options: [
                "00:00-01:00",
                "01:00-02:00",
                "02:00-03:00",
                "03:00-04:00",
                "04:00-05:00",
                "05:00-06:00",
                "06:00-07:00",
                "07:00-08:00",
                "08:00-09:00",
                "09:00-10:00",
                "10:00-11:00",
                "11:00-12:00",
                "12:00-13:00",
                "13:00-14:00",
                "14:00-15:00",
                "15:00-16:00",
                "16:00-17:00",
                "17:00-18:00",
                "18:00-19:00",
                "19:00-20:00",
                "20:00-21:00",
                "21:00-22:00",
                "22:00-23:00",
                "23:00-24:00"
            ].join("\n"),
            default: get_current_hour_slot(),
            onchange: function () {
                frappe.query_report.refresh();
            }
        },
        {
            fieldname: "site",
            label: __("Site"),
            fieldtype: "Link",
            options: "Location",
            reqd: 1,
            onchange: function () {
                frappe.query_report.refresh();
            }
        }
    ],

    onload: function () {
        add_hourly_downtime_styles();
    },

    refresh: function () {
        add_hourly_downtime_styles();

        frappe.after_ajax(function () {
            render_hourly_downtime_html();
        });
    },

    after_datatable_render: function () {
        render_hourly_downtime_html();
    },

    formatter: function (value, row, column, data, default_formatter) {
        value = default_formatter(value, row, column, data);

        if (!data) {
            return value;
        }

        if (column.fieldname === "status") {
            if (data.status_key === "open") {
                return `<span class="hourly-status-open">${value}</span>`;
            }

            return `<span class="hourly-status-available">${value}</span>`;
        }

        return value;
    }
};


function get_current_hour_slot() {
    const now = moment();
    const hour = now.hour();
    const next_hour = hour + 1;

    if (next_hour >= 24) {
        return "23:00-24:00";
    }

    return `${String(hour).padStart(2, "0")}:00-${String(next_hour).padStart(2, "0")}:00`;
}


function add_hourly_downtime_styles() {
    if ($("#hourly-downtime-styles").length) {
        return;
    }

    $("head").append(`
        <style id="hourly-downtime-styles">
            .hourly-downtime-wrapper {
                background: #ffffff;
                border: 1px solid #d9d9d9;
                border-radius: 12px;
                padding: 14px;
                margin: 10px 0 16px 0;
            }

            .hourly-downtime-title {
                font-size: 18px;
                font-weight: 900;
                margin-bottom: 4px;
                text-transform: uppercase;
            }

            .hourly-downtime-subtitle {
                font-size: 13px;
                color: #666;
                margin-bottom: 14px;
            }

            .hourly-downtime-summary {
                display: grid;
                grid-template-columns: repeat(auto-fit, minmax(140px, 1fr));
                gap: 8px;
                margin-bottom: 14px;
            }

            .hourly-summary-box {
                border-radius: 10px;
                border: 1px solid #e5e7eb;
                background: #f8fafc;
                padding: 10px;
                font-size: 12px;
                font-weight: 800;
            }

            .hourly-summary-box strong {
                display: block;
                font-size: 22px;
                margin-top: 4px;
            }

            .hourly-summary-open {
                background: #ffe5e5;
                color: #a8071a;
                border-color: #ff4d4f;
            }

            .hourly-summary-available {
                background: #e6f7e6;
                color: #237804;
                border-color: #52c41a;
            }

            .hourly-available-by-category {
                background: #e6f7e6;
                color: #237804;
                border: 1px solid #52c41a;
                border-radius: 10px;
                padding: 10px;
                margin-bottom: 14px;
            }

            .hourly-available-title {
                font-size: 13px;
                font-weight: 900;
                margin-bottom: 8px;
            }

            .hourly-available-line {
                display: flex;
                justify-content: space-between;
                gap: 12px;
                padding: 4px 0;
                border-bottom: 1px solid rgba(82, 196, 26, 0.25);
                font-size: 13px;
                font-weight: 800;
            }

            .hourly-available-line:last-child {
                border-bottom: none;
            }

            .hourly-category-section {
                margin-top: 12px;
                border: 1px solid #e5e7eb;
                border-radius: 10px;
                overflow: hidden;
            }

            .hourly-category-heading {
                background: #f2f2f2;
                padding: 9px 11px;
                font-weight: 900;
                text-transform: uppercase;
                border-bottom: 1px solid #e5e7eb;
            }

            .hourly-plant-row {
                display: grid;
                grid-template-columns: 34px 120px 1fr;
                gap: 8px;
                align-items: start;
                padding: 8px 11px;
                border-bottom: 1px solid #f0f0f0;
                font-size: 13px;
            }

            .hourly-plant-row:last-child {
                border-bottom: none;
            }

            .hourly-plant-open {
                background: #fff1f0;
                color: #a8071a;
                font-weight: 700;
            }

            .hourly-plant-available {
                background: #f6ffed;
                color: #237804;
            }

            .hourly-plant-status {
                font-size: 16px;
            }

            .hourly-plant-no {
                font-weight: 900;
            }

            .hourly-plant-detail {
                line-height: 1.35;
            }

            .hourly-plant-clickable {
                cursor: pointer;
            }

            .hourly-plant-clickable:hover {
                outline: 2px solid rgba(168, 7, 26, 0.35);
                outline-offset: -2px;
            }

            .hourly-status-open {
                color: #a8071a;
                font-weight: 900;
            }

            .hourly-status-available {
                color: #237804;
                font-weight: 900;
            }
        </style>
    `);
}


function render_hourly_downtime_html() {
    const data = frappe.query_report && frappe.query_report.data ? frappe.query_report.data : [];
    const site = frappe.query_report.get_filter_value("site") || "All Sites";
    const report_date = frappe.query_report.get_filter_value("report_date") || "";
    const hour_slot = frappe.query_report.get_filter_value("hour_slot") || "";

    $(".hourly-downtime-wrapper").remove();

    if (!data.length) {
        $(".report-wrapper").prepend(`
            <div class="hourly-downtime-wrapper">
                <div class="hourly-downtime-title">Hourly Downtime Report</div>
                <div class="hourly-downtime-subtitle">${frappe.utils.escape_html(site)} | ${frappe.utils.escape_html(report_date)} | ${frappe.utils.escape_html(hour_slot)}</div>
                <div>No assets found for this filter.</div>
            </div>
        `);
        return;
    }

    const total = data.length;
    const open = data.filter(row => row.status_key === "open").length;

    const category_order = [
        "ADT",
        "Excavator",
        "Dozer",
        "Water Bowser",
        "Diesel Bowser",
        "Service Truck",
        "FEL",
        "Grader"
    ];

    const grouped = {};

    data.forEach(function (row) {
        const category = row.category_group || row.asset_category || "Uncategorised";

        if (!grouped[category]) {
            grouped[category] = [];
        }

        grouped[category].push(row);
    });

    const available_by_category = {};

    category_order.forEach(function (category) {
        const rows = grouped[category] || [];
        const category_total = rows.length;
        const category_open = rows.filter(function (row) {
            return row.status_key === "open";
        }).length;
        const category_available = category_total - category_open;

        if (category_total > 0) {
            available_by_category[category] = {
                total: category_total,
                available: category_available
            };
        }
    });

    let html = `
        <div class="hourly-downtime-wrapper">
            <div class="hourly-downtime-title">${frappe.utils.escape_html(site)} Fleet Downtime Status</div>
            <div class="hourly-downtime-subtitle">
                Date: ${frappe.utils.escape_html(report_date)} | Hour: ${frappe.utils.escape_html(hour_slot)}
            </div>

            <div class="hourly-downtime-summary">
                <div class="hourly-summary-box">Total Machines<strong>${total}</strong></div>
                <div class="hourly-summary-box hourly-summary-open">Open Downtime<strong>${open}</strong></div>
            </div>

            <div class="hourly-available-by-category">
                <div class="hourly-available-title">Available by Type</div>
                ${category_order.map(function (category) {
                    const item = available_by_category[category];

                    if (!item) {
                        return "";
                    }

                    return `
                        <div class="hourly-available-line">
                            <span>${frappe.utils.escape_html(category)}</span>
                            <strong>${item.available} / ${item.total}</strong>
                        </div>
                    `;
                }).join("")}
            </div>
    `;

    category_order.forEach(function (category) {
        const rows = grouped[category] || [];

        if (!rows.length) {
            return;
        }

        html += `
            <div class="hourly-category-section">
                <div class="hourly-category-heading">${frappe.utils.escape_html(category)}</div>
        `;

        rows.forEach(function (row) {
            const is_open = row.status_key === "open";
            const icon = is_open ? "❌" : "✅";
            const row_class = is_open ? "hourly-plant-open" : "hourly-plant-available";
            const is_clickable = is_open && row.breakdown_docname;
            const clickable_class = is_clickable ? "hourly-plant-clickable" : "";
            const clickable_attr = is_clickable ? `data-breakdown-docname="${frappe.utils.escape_html(row.breakdown_docname)}"` : "";

            let detail = "";

            if (is_open) {
                detail = `
                    ${frappe.utils.escape_html(row.reason || "-")}
                    <br>
                    <small>
                        Open for ${frappe.utils.escape_html(row.open_hours || 0)} hrs at end of hour
                        ${row.start_time ? " | Start: " + frappe.utils.escape_html(row.start_time) : ""}
                    </small>
                `;
            } else {
                detail = frappe.utils.escape_html(row.item_name || "Available");
            }

            html += `
                <div class="hourly-plant-row ${row_class} ${clickable_class}" ${clickable_attr}>
                    <div class="hourly-plant-status">${icon}</div>
                    <div class="hourly-plant-no">${frappe.utils.escape_html(row.plant_no || "")}</div>
                    <div class="hourly-plant-detail">${detail}</div>
                </div>
            `;
        });

        html += `</div>`;
    });

    html += `</div>`;

    $(".report-wrapper").prepend(html);

    $(".hourly-plant-clickable").off("click").on("click", function () {
        const breakdown_docname = $(this).attr("data-breakdown-docname");

        if (breakdown_docname) {
            frappe.set_route("Form", "Plant Breakdown or Maintenance", breakdown_docname);
        }
    });
}

File: engineering/engineering/report/hourly_downtime_report/hourly_downtime_report.py
# Copyright (c) 2026, Isambane Mining (Pty) Ltd
# For license information, please see license.txt

import frappe
from frappe import _
from frappe.utils import getdate, get_datetime, now_datetime, time_diff_in_hours
from datetime import datetime, time, timedelta


CATEGORY_ORDER = {
    "ADT": 1,
    "Excavator": 2,
    "Dozer": 3,
    "Water Bowser": 4,
    "Diesel Bowser": 5,
    "Service Truck": 6,
    "FEL": 7,
    "Grader": 8,
}

# Ignore old dirty open records before this date
STALE_BREAKDOWN_CUTOFF_DATETIME = get_datetime("2026-03-01 00:00:00")


def execute(filters=None):
    filters = frappe._dict(filters or {})
    return get_columns(), get_data(filters)


def get_columns():
    return [
        {"label": _("Status"), "fieldname": "status", "fieldtype": "Data", "width": 120},
        {"label": _("Site"), "fieldname": "site", "fieldtype": "Data", "width": 140},
        {"label": _("Plant No."), "fieldname": "plant_no", "fieldtype": "Data", "width": 120},
        {"label": _("Plant Category"), "fieldname": "category_group", "fieldtype": "Data", "width": 160},
        {"label": _("Plant Model"), "fieldname": "item_name", "fieldtype": "Data", "width": 180},
        {"label": _("Reason"), "fieldname": "reason", "fieldtype": "Small Text", "width": 300},
        {"label": _("Start Time"), "fieldname": "start_time", "fieldtype": "Datetime", "width": 190},
        {"label": _("Resolved Time"), "fieldname": "resolved_time", "fieldtype": "Datetime", "width": 190},
        {"label": _("Open Hours at Hour End"), "fieldname": "open_hours", "fieldtype": "Float", "precision": 2, "width": 170},
        {"label": _("Breakdown Document"), "fieldname": "breakdown_docname", "fieldtype": "Data", "width": 180, "hidden": 1},
    ]


def get_hour_end(filters):
    report_date = getdate(filters.get("report_date")) if filters.get("report_date") else getdate(now_datetime())
    hour_slot = str(filters.get("hour_slot") or "").strip()

    if not hour_slot:
        current_hour = now_datetime().hour
        next_hour = current_hour + 1
        hour_slot = "23:00-24:00" if next_hour >= 24 else f"{current_hour:02d}:00-{next_hour:02d}:00"

    start_text, end_text = hour_slot.split("-")
    end_hour = int(end_text.split(":")[0])

    if end_hour == 24:
        return datetime.combine(report_date + timedelta(days=1), time.min)

    return datetime.combine(report_date, time(end_hour, 0, 0))


def get_category_group(asset_category):
    value = str(asset_category or "").strip().lower()

    if "adt" in value:
        return "ADT"
    if "excavator" in value:
        return "Excavator"
    if "dozer" in value:
        return "Dozer"
    if "water" in value and "bowser" in value:
        return "Water Bowser"
    if "diesel" in value and "bowser" in value:
        return "Diesel Bowser"
    if "service" in value and "truck" in value:
        return "Service Truck"
    if value == "fel" or "front end loader" in value:
        return "FEL"
    if "grader" in value:
        return "Grader"

    return ""


def get_data(filters):
    site = filters.get("site")
    hour_end = get_hour_end(filters)
    hour_start = hour_end - timedelta(hours=1)

    conditions = [
        "a.docstatus = 1",
        "ifnull(a.location, '') != ''",
    ]
    values = {}

    if site:
        conditions.append("a.location = %(site)s")
        values["site"] = site

    assets = frappe.db.sql(
        f"""
        SELECT
            a.name,
            a.location,
            a.asset_category,
            a.item_name
        FROM `tabAsset` a
        WHERE {" AND ".join(conditions)}
        ORDER BY a.location, a.asset_category, a.name
        """,
        values,
        as_dict=True,
    )

    filtered_assets = []

    for asset in assets:
        category_group = get_category_group(
            asset.asset_category
        )

        if not category_group:
            continue

        asset.category_group = category_group
        asset.category_order = CATEGORY_ORDER.get(
            category_group,
            999,
        )
        filtered_assets.append(asset)

    asset_names = [
        asset.name for asset in filtered_assets
    ]

    event_map = get_breakdowns_for_hour(
        asset_names,
        hour_start,
        hour_end,
    )

    data = []

    for asset in filtered_assets:
        breakdown = event_map.get(asset.name)

        if breakdown:
            breakdown_start = get_datetime(
                breakdown.breakdown_start_datetime
            )
            resolved = (
                get_datetime(breakdown.resolved_datetime)
                if breakdown.resolved_datetime
                else None
            )

            overlap_start = max(
                breakdown_start,
                hour_start,
            )
            overlap_end = min(
                resolved or hour_end,
                hour_end,
            )

            hours = round(
                max(
                    (
                        overlap_end - overlap_start
                    ).total_seconds() / 3600,
                    0,
                ),
                2,
            )

            is_open_at_end = (
                not resolved
                or resolved > hour_end
            )

            data.append({
                "status": (
                    "❌ OPEN"
                    if is_open_at_end
                    else "✅ CLOSED"
                ),
                "status_key": (
                    "open"
                    if is_open_at_end
                    else "closed"
                ),
                "site": asset.location,
                "plant_no": asset.name,
                "asset_category": asset.asset_category,
                "category_group": asset.category_group,
                "category_order": asset.category_order,
                "item_name": asset.item_name,
                "reason": breakdown.breakdown_reason or "-",
                "start_time": breakdown.breakdown_start_datetime,
                "resolved_time": breakdown.resolved_datetime,
                "open_hours": hours,
                "breakdown_docname": breakdown.name,
            })
        else:
            data.append({
                "status": "✅ AVAILABLE",
                "status_key": "available",
                "site": asset.location,
                "plant_no": asset.name,
                "asset_category": asset.asset_category,
                "category_group": asset.category_group,
                "category_order": asset.category_order,
                "item_name": asset.item_name,
                "reason": "-",
                "start_time": "",
                "resolved_time": "",
                "open_hours": 0,
                "breakdown_docname": "",
            })

    data.sort(key=lambda row: (
        row.get("category_order") or 999,
        row.get("plant_no") or "",
    ))

    return data


def get_breakdowns_for_hour(
    asset_names,
    hour_start,
    hour_end,
):
    if not asset_names:
        return {}

    rows = frappe.db.sql(
        """
        SELECT
            name,
            asset_name,
            breakdown_reason,
            breakdown_start_datetime,
            resolved_datetime
        FROM `tabPlant Breakdown or Maintenance`
        WHERE asset_name IN %(asset_names)s
          AND breakdown_start_datetime < %(hour_end)s
          AND (
              resolved_datetime IS NULL
              OR resolved_datetime > %(hour_start)s
          )
        ORDER BY
            asset_name,
            breakdown_start_datetime DESC,
            modified DESC
        """,
        {
            "asset_names": tuple(asset_names),
            "hour_start": hour_start,
            "hour_end": hour_end,
        },
        as_dict=True,
    )

    event_map = {}

    for row in rows:
        start_time = get_datetime(
            row.breakdown_start_datetime
        )
        resolved = (
            get_datetime(row.resolved_datetime)
            if row.resolved_datetime
            else None
        )

        if (
            not resolved
            and start_time
            < STALE_BREAKDOWN_CUTOFF_DATETIME
        ):
            continue

        current = event_map.get(row.asset_name)

        # Prefer a breakdown open at the period end.
        is_open_at_end = (
            not resolved
            or resolved > hour_end
        )

        if (
            current is None
            or (
                is_open_at_end
                and current.get("_status") != "open"
            )
        ):
            row["_status"] = (
                "open" if is_open_at_end else "closed"
            )
            event_map[row.asset_name] = row

    return event_map
